### PR TITLE
Aggregator PoC

### LINKED ISSUE #674: End-to-end Aggregator proof-of-concept
We need to implement an end-to-end Aggregator proof-of-concept that includes the following components:
- Client application
    - Sends data to the Aggregator
    - In general data should be represented as a [Sparse Vector](https://en.wikipedia.org/wiki/Sparse_matrix) what in this particular example means a dictionary.
- Aggregator instance
    - Aggregates data from clients (implemented in #626)
    - Has multiple aggregation buckets for aggregating samples. Each bucket has it's own dedicated sample threshold (minimal number of aggregated samples)
    - Waits for the threshold and sends data to the Backend server using gRPC
- Backend server
    - Collects aggregated data from Aggregator instances
    - Provides a UI to show the data timeline (e.g. [Prometheus](https://prometheus.io/))
- Scripts to run all this infrastructure in Google Cloud Platform (GCP)

_Optional:_
Use native GCP load balancing to balance client requests between multiple Aggregator instances

### PR DESCRIPTION
**Pull request by ipetr0v**
This change adds an Aggregator PoC application that works with Sparse Vectors.
Previous version of the Aggregator remains as `aggregator_test`.

Fixes #674

# Checklist

- [x] Pull request affects core Oak functionality (e.g. runtime, SDK, ABI)
  - [x] I have written tests that cover the code changes.
  - [x] I have checked that these tests are run by [Cloudbuild](cloudbuild.yaml)
  - [ ] I have updated [documentation](docs/) accordingly.
  - [ ] I have raised an [issue](https://github.com/project-oak/oak/issues) to
        cover any TODOs and/or unfinished work.
- [ ] Pull request includes prototype/experimental work that is under
      construction.

### CODE REVIEW

**Comment by ipetr0v** [comment]
@tiziano88 should we keep the previous version of the Aggregator (that works with regular vectors) as a `aggregator_test` (also in the `examples/aggregator/` directory) ?

File: examples/aggregator/proto/aggregator.proto [hunk 1]
```diff
@@ -20,11 +20,18 @@ package oak.examples.aggregator;
 
 import "google/protobuf/empty.proto";
 
-message Vector {
-  repeated uint64 items = 1;
+message SerializedSparseVector {
```
Comment (tiziano88) [inline examples/aggregator/proto/aggregator.proto]: Add some documentation.

File: examples/aggregator/proto/aggregator.proto [hunk 2]
```diff
@@ -20,11 +20,18 @@ package oak.examples.aggregator;
...
+  repeated uint32 indices = 1;
+  repeated float values = 2;
+}
+
+// Message sent by clients.
+message Sample {
+  // Data with different Bucket IDs should not be mixed (aggregated).
```
Comment (tiziano88) [inline examples/aggregator/proto/aggregator.proto]: Clarify this comment, it is not clear whether it is asking clients to do something (not sure what), or whether it is describing a guarantee offered by the server.

File: examples/aggregator/module/rust/src/lib.rs [hunk 3]
```diff
@@ -22,76 +22,117 @@
...
+const SAMPLE_THRESHOLD: u64 = 2;
 
-oak::entrypoint!(oak_main => {
-    oak::logger::init_default();
-    let node = AggregatorNode {
-        aggregator: ThresholdAggregator::<Vector>::new(SAMPLE_THRESHOLD),
-    };
-    Dispatcher::new(node)
-});
+type ThresholdAggregatorMap = HashMap<String, Option<ThresholdAggregator<SparseVector>>>;
```
Comment (tiziano88) [inline examples/aggregator/module/rust/src/lib.rs]: This map seems to obfuscate things instead of making them clearer. I would just inline the extended type below. In any case, could you clarify why there is an `Option` on the RHS?

File: examples/aggregator/module/rust/src/lib.rs [hunk 4]
```diff
@@ -22,76 +22,117 @@
...
-    aggregator: ThresholdAggregator<Vector>,
+    fn send_aggregated_data(&self, bucket: String, svec: SparseVector) {
+        debug!(
+            "Sending aggregated data: bucket {}, sparse vector: {:?}",
+            bucket, svec
+        );
+
+        match oak::grpc::client::Client::new("grpc-client", "ignored").map(AggregatorClient) {
+            Some(grpc_client) => {
+                // Currently the Node panics if it couldn't send a gRPC request.
```
Comment (tiziano88) [inline examples/aggregator/module/rust/src/lib.rs]: That seems extreme, it should be possible to just log?

File: examples/aggregator/module/rust/src/lib.rs [hunk 5]
```diff
@@ -22,76 +22,117 @@
...
+                grpc_client
+                    .submit_sample(Sample {
+                        bucket: bucket,
+                        data: ::protobuf::SingularPtrField::some(SerializedSparseVector::from(
```
Comment (tiziano88) [inline examples/aggregator/module/rust/src/lib.rs]: Is it possible to use `.into()` here?

File: examples/aggregator/module/rust/src/lib.rs [hunk 6]
```diff
@@ -22,76 +22,117 @@
...
-/// Oak Node that collects aggregated data.
-pub struct AggregatorNode {
-    aggregator: ThresholdAggregator<Vector>,
+    fn send_aggregated_data(&self, bucket: String, svec: SparseVector) {
+        debug!(
+            "Sending aggregated data: bucket {}, sparse vector: {:?}",
+            bucket, svec
+        );
+
+        match oak::grpc::client::Client::new("grpc-client", "ignored").map(AggregatorClient) {
```
Comment (tiziano88) [inline examples/aggregator/module/rust/src/lib.rs]: What's `"ignored"`?

File: examples/aggregator/module/rust/src/lib.rs [hunk 7]
```diff
@@ -22,76 +22,117 @@
...
-            )
-        });
-        info!("Returning value: {:?}", value);
-        value
+    fn submit_sample(&mut self, sample: Sample) -> grpc::Result<Empty> {
+        match sample.data.into_option() {
+            Some(data) => match SparseVector::try_from(data) {
+                Ok(svec) => {
+                    debug!(
+                        "Submitted data: bucket {}, sparse vector: {:?}",
```
Comment (tiziano88) [inline examples/aggregator/module/rust/src/lib.rs]: Should this be `Received data` rather?

File: examples/aggregator/module/rust/src/lib.rs [hunk 8]
```diff
@@ -22,76 +22,117 @@
...
-                .iter()
-                .zip_longest(other.items.iter())
-                .map(|p| match p {
-                    Both(l, r) => *l + *r,
-                    Left(l) => *l,
-                    Right(r) => *r,
-                })
-                .collect(),
-            ..Default::default()
+    fn submit_sparse_vector(&mut self, bucket: String, svec: &SparseVector) -> Result<(), String> {
```
Comment (tiziano88) [inline examples/aggregator/module/rust/src/lib.rs]: This method name seems misleading; `submit` makes sense as the gRPC name from the client point of view, but having an internal method called `submit` that does not actually submit anything does not seem very accurate.

But I think this actually comes from this name:

https://github.com/project-oak/oak/blob/0bb76ad2b58e1a26c551e3805cf6089c846308e0/examples/aggregator/module/rust/src/aggregator.rs#L48

Perhaps it should be renamed to something else, but I don't have a good suggestion. A few random things that come to mind but do not seem ideal: `add`, `contribute`, `merge`.

@project-oak/core any suggestions?

File: examples/aggregator/module/rust/src/data.rs [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+//
+
+use crate::proto::aggregator::SerializedSparseVector;
+use aggregator_common::Monoid;
+use std::collections::HashMap;
+use std::convert::{From, TryFrom};
+
+// Sparce Vectors are stored as Hash Maps.
+#[derive(Clone, Debug)]
+pub struct SparseVector {
```
[the remark this answers is not in the document]
Comment (ipetr0v) [reply examples/aggregator/module/rust/src/data.rs]: I tried, but it's not possible to define traits for structs from a different crate

File: examples/aggregator/module/rust/src/data.rs [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+    map: HashMap<u32, f32>,
+}
+
+// Deserializes a Protobuf message into an internal Sparce Vector implementation.
+// If a Protobuf message has duplicated indices, returns an `Error`.
+impl TryFrom<SerializedSparseVector> for SparseVector {
```
[the remark this answers is not in the document]
Comment (ipetr0v) [reply examples/aggregator/module/rust/src/data.rs]: If I implement a `TryFrom` trait - I cannot use reference, since the trait defines a `fn try_from(value: T) -> Result<Self, Self::Error>` without a reference
Or I just should create a new trait?

File: examples/aggregator/README.md [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+### Backend
+
+Backend server for the Aggregator example.
+
+Backend is used in tests/experiments and is represented as a gRPC server that
+listens for samples provided by the Aggregator and prints them into the standard
+output.
+
+Build and run the Backend with the following command:
+```bash
```
Comment (tiziano88) [inline examples/aggregator/README.md]: (nit) add extra lines around code blocks.

File: examples/aggregator/README.md [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+cargo run --release --manifest-path=examples/aggregator/backend/Cargo.toml
```
Comment (tiziano88) [inline examples/aggregator/README.md]: Does this work?

```suggestion
cargo run --release --package=aggregator_backend
```

File: examples/aggregator/README.md [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+```
+
+Backend code is in the `backend` directory.
+
+### Client
+
+Simple client that connects to the Aggregator and sends a data sample with gRPC.
```
Comment (tiziano88) [inline examples/aggregator/README.md]:
```suggestion
Simple client that connects to the Aggregator and sends a data sample via gRPC.
```

File: examples/aggregator/module/rust/src/data.rs [hunk 14]
```diff
@@ -60,15 +60,18 @@ impl From<SparseVector> for SerializedSparseVector {
...
     fn combine(&self, other: &Self) -> Self {
-        other.map.iter().fold(self.clone(), |mut svec, (&i, &v)| {
-            *svec.map.entry(i).or_insert(v) += v;
-            svec
-        })
+        other
+            .entries
+            .iter()
+            .fold(self.clone(), |mut svec, (&i, &v)| {
+                *svec.entries.entry(i).or_insert(v) += v;
```
Comment (tiziano88) [inline examples/aggregator/module/rust/src/data.rs]: Doesn't this cause v to be calculated twice, if the entry was empty? I expect it should be

```suggestion
                *svec.entries.entry(i).or_insert(0) += v;
```

BTW, perhaps some unit tests would be useful here :)

File: examples/aggregator/common/src/lib.rs [hunk 15]
```diff
@@ -59,4 +61,14 @@ impl<T: Monoid> ThresholdAggregator<T> {
             None
         }
     }
+
+    /// Deletes current aggregated value and returns it if the number of collected samples is at
+    /// least `sample_threshold`.
+    pub fn take(&mut self) -> Option<T> {
+        if self.sample_count >= self.sample_threshold {
+            Some(replace(&mut self.current_value, Monoid::identity()))
```
[the remark this answers is not in the document]
Comment (tiziano88) [reply examples/aggregator/common/src/lib.rs]: I think some unit test should have caught this, could you add some?

File: examples/aggregator/module/rust/src/lib.rs [hunk 16]
```diff
@@ -38,30 +38,32 @@ use std::collections::HashMap;
...
 pub struct AggregatorNode {
-    aggregators: ThresholdAggregatorMap,
+    aggregators: HashMap<String, Option<ThresholdAggregator<SparseVector>>>,
 }
 
 impl AggregatorNode {
     fn new() -> Self {
         AggregatorNode {
-            aggregators: ThresholdAggregatorMap::new(),
+            aggregators: HashMap::<String, Option<ThresholdAggregator<SparseVector>>>::new(),
```
Comment (tiziano88) [inline examples/aggregator/module/rust/src/lib.rs]: Is the explicit type necessary? I thought `HashMap::new()` should be enough?

File: examples/aggregator/module/rust/src/lib.rs [hunk 17]
```diff
@@ -38,30 +38,32 @@ use std::collections::HashMap;
...
         }
     }
 
-    fn submit_sparse_vector(&mut self, bucket: String, svec: &SparseVector) -> Result<(), String> {
+    fn try_report_aggregation(&mut self, bucket: String, svec: &SparseVector) -> Result<(), String> {
```
Comment (tiziano88) [inline examples/aggregator/module/rust/src/lib.rs]: Add comment.

File: examples/aggregator/module/rust/src/lib.rs [hunk 18]
```diff
@@ -75,24 +77,22 @@ impl AggregatorNode {
         Ok(())
     }
 
-    fn send_aggregated_data(&self, bucket: String, svec: SparseVector) {
+    fn report(&self, bucket: String, svec: SparseVector) {
```
Comment (tiziano88) [inline examples/aggregator/module/rust/src/lib.rs]: Add comment.

Comment (tiziano88) [inline examples/aggregator/module/rust/src/lib.rs]: Just curious, why is this not returning a `Result` too? The callers already handle logging of errors (line 114).

File: examples/aggregator/module/rust/src/lib.rs [hunk 19]
```diff
@@ -106,25 +106,22 @@ impl Aggregator for AggregatorNode {
...
                     );
-                    match self.submit_sparse_vector(sample.bucket, &svec) {
-                        Ok(_) => Ok(Empty::new()),
-                        Err(err) => {
-                            error!("Sparse Vector submission error: {}", err);
-                            Err(grpc::build_status(grpc::Code::INVALID_ARGUMENT, &err))
-                        }
+                    if let Err(err) = self.try_report_aggregation(sample.bucket, &svec) {
+                        let err = format!("Backend report error: {}", err);
+                        error!("{:?}", err);
```
Comment (tiziano88) [inline examples/aggregator/module/rust/src/lib.rs]: Why this double format, first via `format` and then via `error`?

File: examples/aggregator/module/rust/src/lib.rs (hunk 6 above)
Comment (tiziano88) [reply examples/aggregator/module/rust/src/lib.rs]: I'm still not sure what this is used for. How can a gRPC pseudo-node have an entrypoint name? Perhaps @daviddrysdale can clarify.

File: examples/aggregator/module/rust/src/data.rs [hunk 20]
```diff
@@ -0,0 +1,74 @@
...
+impl TryFrom<SerializedSparseVector> for SparseVector {
+    type Error = String;
+
+    fn try_from(src: SerializedSparseVector) -> Result<Self, Self::Error> {
+        src.indices.iter().zip(src.values.iter()).try_fold(
+            SparseVector::identity(),
+            |mut svec, (&i, &v)| match svec.map.get(&i) {
+                Some(_) => Err(format!("Duplicated index: {}", i)),
+                None => Ok({
+                    svec.map.insert(i, v);
```
[the remark this answers is not in the document]
Comment (ipetr0v) [reply examples/aggregator/module/rust/src/data.rs]: But even if I will use `entry` instead of `insert` I will still need to reborrow the map in the `fold`.
Or you mean not using `fold` at all?

File: examples/aggregator/module/rust/src/data.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (tiziano88) [reply examples/aggregator/module/rust/src/data.rs]: Does `impl TryFrom<&SerializedSparseVector>` not work?

File: examples/aggregator/module/rust/src/data.rs [hunk 21]
```diff
@@ -0,0 +1,77 @@
...
+use std::collections::HashMap;
+use std::convert::{From, TryFrom};
+
+// Sparse Vectors are stored as Hash Maps.
+#[derive(Clone, Debug)]
+pub struct SparseVector {
+    entries: HashMap<u32, f32>,
+}
+
+// Deserializes a Protobuf message into an internal Sparce Vector implementation.
```
Comment (tiziano88) [inline examples/aggregator/module/rust/src/data.rs]: s/Sparce/Sparse/

File: examples/aggregator/module/rust/src/data.rs [hunk 22]
```diff
@@ -0,0 +1,77 @@
...
+                None => Ok({
+                    svec.entries.insert(i, v);
+                    svec
+                }),
+            },
+        )
+    }
+}
+
+// Serializes an internal Sparce Vector implementation into a Protobuf message.
```
Comment (tiziano88) [inline examples/aggregator/module/rust/src/data.rs]: s/Sparce/Sparse/

File: examples/aggregator/module/rust/src/data.rs (hunk 20 above)
[the remark this answers is not in the document]
Comment (tiziano88) [reply examples/aggregator/module/rust/src/data.rs]: I meant `entry` instead of `get` actually, and keep the `fold`. I'm not even sure if it works though, let me know if you try it out.

[the remark this answers is not in the document]
Comment (ipetr0v) [reply examples/aggregator/module/rust/src/data.rs]: The problem with `entry` is that I still need to return error from the function inside `try_fold`.
Also, I'm not sure, why `Copy` is necessary - `svec` in the `|mut svec, (&i, &v)|` is just moved.

File: examples/aggregator/module/rust/Cargo.toml [hunk 23]
```diff
@@ -8,8 +8,10 @@ edition = "2018"
 crate-type = ["cdylib"]
 
 [dependencies]
+aggregator_common = { path = "../../common" }
 itertools = "*"
 log = "*"
+maplit = "*"
```
Comment (tiziano88) [inline examples/aggregator/module/rust/Cargo.toml]: Isn't this a dev dependency?

File: examples/aggregator/module/rust/src/tests.rs [hunk 24]
```diff
@@ -46,31 +58,103 @@ fn test_aggregator() {
...
+        ),
+        Err(_)
+    );
 
     runtime.stop();
 }
+
+fn check_combine(src1: HashMap<u32, f32>, src2: HashMap<u32, f32>, dst: HashMap<u32, f32>) -> bool {
+    let src = SparseVector::new(src1).combine(&SparseVector::new(src2));
+    src == SparseVector::new(dst)
```
Comment (tiziano88) [inline examples/aggregator/module/rust/src/tests.rs]: May as well do the assert here :) otherwise, if you had more things to check, it would not be feasible to return everything, and expect the caller would just assert all of them.

An alternative would be to write a table-driven test, e.g.

```rust
#[test]
fn test_combine() {
  type Map = HashMap<u32, f32>;
  struct Test {
    in_0: Map,
    in_1: Map,
    out: Map,
  }
  let tests = vec![
    Test {
      in_0: hashmap! {},
      in_1: hashmap! { 1 => 10.0 },
      out: hashmap! { 1 => 10.0 },
    },
    ...
  ];
  for test in tests {
    ...
  }
}
```

and also below you could use the same approach for other tests.

WDYT?

File: examples/aggregator/module/rust/src/lib.rs (hunk 6 above)
Comment (daviddrysdale) [reply examples/aggregator/module/rust/src/lib.rs]: A pseudo-Node doesn't have an entrypoint name, but the same `node_create` host function is used for creating both Wasm Nodes (which need a module/config name and entrypoint name) and pseudo-Nodes (which just need a config name); the `"ignored"` placeholder is a reminder.

File: examples/aggregator/common/src/lib.rs (hunk 15 above)
[the remark this answers is not in the document]
Comment (tiziano88) [reply examples/aggregator/common/src/lib.rs]: (I still think that some **unit** tests for this package would be useful)

File: examples/aggregator/module/rust/src/data.rs (hunk 20 above)
[the remark this answers is not in the document]
Comment (tiziano88) [reply examples/aggregator/module/rust/src/data.rs]: To clarify: I am not talking about `svec`; I think you can currently re-borrow map inside match, because when you do `match sve.map.get(&i)`, this value is immediately copied, and therefore the borrow from the `get` does not last inside the body of the `match`. If it did (i.e. if the type was not `Copy`, then I think you would not be able to call `insert` here. But I haven't actually tried :)

File: examples/aggregator/module/rust/src/lib.rs (hunk 6 above)
Comment (tiziano88) [reply examples/aggregator/module/rust/src/lib.rs]: Indeed, so why expose the parameter in `Client::new` at all? AFAICT no one would even pass anything useful via it. It will still use `node_create` internally of course, but the placeholder will be hidden inside the implementation of `new`, where it can also be documented.

Comment (daviddrysdale) [reply examples/aggregator/module/rust/src/lib.rs]: It's needed because another Wasm node may provide a gRPC service, and the `grpc::client:Client` code works with either a remote (non-Oak) gRPC service (via the gRPC client pseudo-Node), or a local gRPC service (provided by another Oak Node in the same Application).  

TL;DR : [It's required here](https://github.com/project-oak/oak/blob/master/examples/hello_world/module/rust/src/lib.rs#L29)

File: examples/aggregator/module/rust/src/data.rs (hunk 20 above)
[the remark this answers is not in the document]
Comment (ipetr0v) [reply examples/aggregator/module/rust/src/data.rs]: Done

[the remark this answers is not in the document]
Comment (ipetr0v) [reply examples/aggregator/module/rust/src/data.rs]: But I'm returning `Ok(svec)` from the `Entry::Vacant(entry)`.
Does it mean that I'm still reborrowing `svec` mutably?

[the remark this answers is not in the document]
Comment (tiziano88) [reply examples/aggregator/module/rust/src/data.rs]: I haven't actually tried any of the stuff I mentioned here, it's just something that occurred to me :) and I was not necessarily suggesting anything needs to change in the code. I was mostly hoping you could try out various options (e.g. Copy vs non-Copy, entry vs get) and let the rest of us know :)

[the remark this answers is not in the document]
Comment (tiziano88) [reply examples/aggregator/module/rust/src/data.rs]: It may well be that the only way to make this work for a non-Copy type is to just use a `for` loop and `entry` to access an individual entry from the map. Let me know if you end up trying that out, separately. I think the code here looks fine anyways, no need to change it further, unless you want to.
